File: modules/books-widget/src/BooksWidgetModule.ts
import { requireOptionalNativeModule } from 'expo';

/**
 * The native bridge to the Android home-screen widget. Backed by Kotlin
 * (`BooksWidgetModule.kt`); present only on Android. `requireOptionalNativeModule`
 * returns `null` on iOS/web/Expo Go so callers can no-op without crashing.
 */
export interface BooksWidgetNativeModule {
  /** Persist the JSON widget snapshot the Glance widget renders from. */
  writeSnapshot(json: string): Promise<void>;
  /** Re-render all live widget instances from the current snapshot. */
  refreshWidget(): Promise<void>;
  /**
   * Return and clear the queue of "marked read from the widget" dates as a JSON
   * array string (e.g. `["2026-06-22"]`). Used to reconcile widget taps into the
   * database on app launch.
   */
  getPendingMarks(): Promise<string>;
}

export default requireOptionalNativeModule<BooksWidgetNativeModule>('BooksWidget');

File: modules/books-widget/android/src/main/java/dk/jonathanb/books/widget/BooksWidgetModule.kt
package dk.jonathanb.books.widget

import kotlinx.coroutines.runBlocking
import androidx.glance.appwidget.updateAll
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * JS-callable bridge for the Android home-screen widget. Writes the snapshot the
 * Glance widget renders from, triggers re-renders, and hands back the queue of
 * marks tapped on the widget so the app can reconcile them into SQLite.
 *
 * Android-only; on other platforms `requireOptionalNativeModule('BooksWidget')`
 * returns null and the JS service no-ops.
 */
class BooksWidgetModule : Module() {
  private val context
    get() = requireNotNull(appContext.reactContext) { "React context unavailable" }

  override fun definition() = ModuleDefinition {
    Name("BooksWidget")

    AsyncFunction("writeSnapshot") { json: String ->
      WidgetPrefs.prefs(context).edit().putString(WidgetPrefs.KEY_SNAPSHOT, json).commit()
    }

    AsyncFunction("refreshWidget") {
      // updateAll re-runs provideGlance for every live instance, re-reading prefs.
      runBlocking { BooksWidget().updateAll(context) }
    }

    AsyncFunction("getPendingMarks") {
      WidgetPrefs.takePendingMarks(context)
    }
  }
}
